drivers/hmis_csv_importer/app/models/hmis_csv_importer/importer/importer.rb

```diff
@@ -101,22 +108,224 @@ def resume!
+    # to be used in import threshold checks before committing data changes.
+    #
+    # @return [void]
+    #
+    private def precalculate_change_counts
+      # This makes an estimate of the changes that will occur as it needs to be done before the
+      # actual processing so that it can pause the import if necessary.
+      # NOTE: as written this causes 2 additional, potentially slow queries
+      importable_files.map do |file, klass|
+        incoming_data = klass.incoming_data(importer_log_id: importer_log.id).
+          pluck(klass.hud_key).to_set
+
+        to_add = (incoming_data - existing_hud_keys(klass)).count
+        to_remove = (existing_hud_keys(klass) - incoming_data).count
+        # If we never delete, just pretend it will be 0
+        to_remove = 0 if klass.prevent_import_deletions?
+        importer_log.summary[file]['added'] = to_add
+        importer_log.summary[file]['removed'] = to_remove
+      end
+    end
 
-      loader_errors.positive? || db_errors.count.positive? || validation_errors.count.positive?
+    ##
+    # Retrieves the set of existing HUD keys from the warehouse for a given class.
+    #
+    # This method queries the warehouse for records that match the data source,
+    # involved projects, and date range. It then extracts and returns the HUD keys
+    # as a set for efficient lookup and comparison.
+    #
+    # @param klass [Class] The class representing the data model being queried.
+    # @return [Set<String>] A set of existing HUD keys for the given class.
+    #
+    memoize private def existing_hud_keys(klass)
+      klass.existing_data(
+        data_source_id: data_source.id,
+        project_ids: involved_project_ids,
+        date_range: date_range,
+      ).pluck(klass.hud_key).to_set
+    end
+
+    ##
+    # Collects and aggregates changes in record counts during the import process.
+    #
+    # This method calculates changes in records (additions and removals) for each importable file.
+    # It determines the total number of changes by comparing additions and removals and associates
+    # the changes with the corresponding file. Additionally, it retrieves the total record count
+    # for of existing data that matches each file to provide more context about the scale of the changes.
+    #
+    # @return [Hash{String => Hash{Symbol => Integer}}] A hash where keys are filenames, and values
+    #   are hashes containing:
+    #   - `:change_count` (Integer): The absolute value of the difference between additions and removals.
+    #   - `:total_count` (Integer): The total number of records for the corresponding file.
+    #
+    private def change_counts
+      importer_log.summary.map do |file, data|
+        klass = importable_files[file]
+        to_add = data['added']
+        to_remove = data['removed']
+        [
+          file,
+          {
+            change_count: (to_add - to_remove).abs,
```

Comment:
I am probably not understanding but I'm unsure about this method of calculating changes. If I have 100 records, I remove all 100 and add another 100, I effectively have 0 changes even though I have replaced the entire record set? That seems expected.

I think summing them would be more intuitive, no? Or maybe max(inserts, deletions)

Reply:
Maybe the max of the two.  Generally speaking it's ok if we "replace" them.  What we're trying to catch is where the overall magnitude of the change is large.  If the sending HMIS is just renumbering the HUD Keys, but the overall change will be small, we probably don't care.